src/smpp_socket.erl

```diff
@@ -1000,3 +1033,69 @@ check_in_flight(#{in_flight := InFlight, in_flight_limit := Limit} = State)
     {ok, State};
 check_in_flight(State) ->
     {error, {in_flight_limit_cooldown, State, [postpone]}}.
+
+%%%-------------------------------------------------------------------
+%%% ESME back pressure
+%%%-------------------------------------------------------------------
+-define(PT_MAX_AWAIT_REQS(Id), {?MODULE, Id, max_await_reqs}).
+-define(PT_AWAIT_REQS_COUNTER(Id), {?MODULE, Id, await_reqs_counter}).
+
+-spec init_req_counter(state()) -> state().
+init_req_counter(#{max_await_reqs := undefined} = State) ->
+    State;
+init_req_counter(#{max_await_reqs := MaxAwaitReqs, id := Id} = State) ->
+    %% Back pressure here does not rely on exact counter value by design.
+    %% I.e. a bit more than MaxAwaitReqs can be queued due to a natural race
+    %% but then back pressure will kick in anyway. That is why 'write_concurrency'
+    %% counters option fits well here.
+    CounterRef = counters:new(1, [write_concurrency]),
+    ok = persistent_term:put(?PT_AWAIT_REQS_COUNTER(Id), CounterRef),
+    ok = persistent_term:put(?PT_MAX_AWAIT_REQS(Id), MaxAwaitReqs),
+    State#{await_reqs_counter => CounterRef}.
+
+-spec enqueue_req(gen_statem:server_ref()) -> ok | error.
+enqueue_req(Ref) ->
+    enqueue_req(Ref, undefined).
+
+-spec enqueue_req(gen_statem:server_ref(), millisecs() | undefined) -> ok | error.
+enqueue_req(Pid, _Timeout) when is_pid(Pid) ->
+    %% TODO: add back pressure support for direct Pid invocations.
+    ok;
+enqueue_req(Name, Timeout) ->
+    Id = get_id_by_name(Name),
+    case persistent_term:get(?PT_MAX_AWAIT_REQS(Id), undefined) of
+        undefined ->
+            ok;
+        MaxAwaitReqs ->
+            CounterRef = persistent_term:get(?PT_AWAIT_REQS_COUNTER(Id)),
+            case counters:get(CounterRef, 1) of
```

Comment:
there is possible race condition between 'get' and subsequent 'add' operations, so, second order errors (false positives) in 'enqueue' are still possible. Are we okay with this?

Reply:
As mentioned in the comment to init_